File: backend/src/main/java/com/back/domain/product/entity/Product.java
package com.back.domain.product.entity;

import com.back.domain.order.entity.OrderItem;
import com.back.domain.product.enums.StockStatus;
import com.back.domain.product.exception.ProductErrorCode;
import com.back.domain.product.exception.ProductException;
import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Getter
@NoArgsConstructor(access = lombok.AccessLevel.PROTECTED)
@Table(name = "product")
@EntityListeners(AuditingEntityListener.class)
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "product_id")
    private Long id;

    @Column(name = "product_name", nullable = false)
    private String name;

    @Column(name = "product_price", nullable = false)
    private int price;

    @Column(name = "product_description")
    @Lob
    private String description;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime editedAt;

    @Column(name = "image_path")
    private String imagePath;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<OrderItem> orderItems;

    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JoinColumn(name = "stock_id")
    private Stock stock;

    @Builder
    public Product(String name, int price, String description, String imagePath) {
        this.name = name;
        this.price = price;
        this.description = description;
        this.imagePath = imagePath;
    }

    public void update(String name, int price, String description, String imagePath, Stock stock) {
        this.name = name;
        this.price = price;
        this.description = description;
        this.imagePath = imagePath;
        this.stock = stock;
    }

    public void setStock(Stock stock) {
        if (this.stock != null) {
            throw new ProductException(ProductErrorCode.PRODUCT_OUT_OF_STOCK);
        }
        this.stock = stock;
    }

    public void increaseStock(int count) {
        if (stock == null) {
            stock = new Stock(count, StockStatus.IN_STOCK, this);
        } else {
            stock.increaseCount(count);
        }
    }

    public void decreaseStock(@Min(1) @Max(999) int quantity) {
        if (stock == null || stock.getQuantity() < quantity) {
            throw new ProductException(ProductErrorCode.PRODUCT_NOT_ENOUGH_STOCK);
        }
        stock.updateQuantity(stock.getQuantity() - quantity);
        if (stock.getQuantity() <= 0) {
            stock.updateStockStatus(StockStatus.OUT_OF_STOCK);
        }
    }
}

File: frontend/src/types/dev/product.ts
// API 응답 타입
export interface ProductApiResponse {
    success: boolean;
    code: string;
    message: string;
    content: ProductApiItem[];
}

export interface ProductApiItem {
    id: number;
    name: string;
    price: number;
    description: string;
    createdAt: string;
    editedAt: string;
    imagePath: string;
    stockDto: {
        quantity: number;
        stockStatus: string;
    };
}

// 프론트엔드에서 사용하는 Product 타입
export interface Product {
    id: number;
    name: string;
    price: number;
    imagePath: string;
    description: string;
    createdAt: string;
    editedAt: string;
    stock: number;
}

// API 응답을 프론트엔드 타입으로 변환하는 함수
export function mapProductApiToProduct(apiItem: ProductApiItem): Product {
    return {
        id: apiItem.id,
        name: apiItem.name,
        price: apiItem.price,
        imagePath: apiItem.imagePath,
        description: apiItem.description,
        createdAt: apiItem.createdAt,
        editedAt: apiItem.editedAt,
        stock: apiItem.stockDto?.quantity || 0
    };
}

// API 응답 배열을 프론트엔드 타입 배열로 변환하는 함수
export function mapProductApiArrayToProducts(apiItems: ProductApiItem[]): Product[] {
    return apiItems.map(mapProductApiToProduct);
}
